### PR TITLE
bug/2584_service_path_as_http_header_in_notifications

### PR DESCRIPTION
**Pull request by kzangeli**
Fixed issue #2584.

### CODE REVIEW

File: test/functionalTest/cases/0673_xauth_token/auth_token_from_subscribe_request.test [hunk 1]
```diff
@@ -192,6 +192,8 @@ Date: REGEX(.*)
 ========================================================================================
 POST http://127.0.0.1:REGEX(\d+)/notify
 Content-Length: 459
+Fiware-Servicepath: /
+Fiware-Servicepath: /
```
Comment (fgalan) [inline test/functionalTest/cases/0673_xauth_token/auth_token_from_subscribe_request.test]: Duplicated

File: test/functionalTest/cases/0673_xauth_token/auth_token_from_subscribe_request.test [hunk 2]
```diff
@@ -251,6 +253,8 @@ Date: REGEX(.*)
 ==================================================================================================
 POST http://127.0.0.1:REGEX(\d+)/notify
 Content-Length: 459
+Fiware-Servicepath: /
+Fiware-Servicepath: /
```
Comment (fgalan) [inline test/functionalTest/cases/0673_xauth_token/auth_token_from_subscribe_request.test]: Duplicated

File: test/functionalTest/cases/0673_xauth_token/auth_token_three_steps.test [hunk 3]
```diff
@@ -237,6 +237,8 @@ Date: REGEX(.*)
 ========================================================================================
 POST http://127.0.0.1:REGEX(\d+)/notify
 Content-Length: 459
+Fiware-Servicepath: /
+Fiware-Servicepath: /
```
Comment (fgalan) [inline test/functionalTest/cases/0673_xauth_token/auth_token_three_steps.test]: Duplicated

File: test/functionalTest/cases/0673_xauth_token/auth_token_three_steps.test [hunk 4]
```diff
@@ -314,6 +316,8 @@ Date: REGEX(.*)
 ================================================================================
 POST http://127.0.0.1:REGEX(\d+)/notify
 Content-Length: 460
+Fiware-Servicepath: /
+Fiware-Servicepath: /
```
Comment (fgalan) [inline test/functionalTest/cases/0673_xauth_token/auth_token_three_steps.test]: Duplicated

File: test/functionalTest/cases/0673_xauth_token/subscription.test [hunk 5]
```diff
@@ -142,6 +142,8 @@ Date: REGEX(.*)
 =============================================================
 POST http://127.0.0.1:REGEX(\d+)/notify
 Content-Length: 459
+Fiware-Servicepath: /
+Fiware-Servicepath: /
```
Comment (fgalan) [inline test/functionalTest/cases/0673_xauth_token/subscription.test]: Duplicated

File: test/functionalTest/cases/0715_forward_xauth_token_to_cprs/propagate_xauth_token_for_update_and_query_standard_ops.test [hunk 6]
```diff
@@ -244,6 +244,8 @@ Date: REGEX(.*)
 ===================================================
 POST http://localhost:REGEX(\d+)/v1/updateContext
 Content-Length: 274
+Fiware-Servicepath: /
+Fiware-Servicepath: /
```
Comment (fgalan) [inline test/functionalTest/cases/0715_forward_xauth_token_to_cprs/propagate_xauth_token_for_update_and_query_standard_ops.test]: Dup

File: test/functionalTest/cases/0715_forward_xauth_token_to_cprs/propagate_xauth_token_for_update_and_query_standard_ops.test [hunk 7]
```diff
@@ -271,6 +273,8 @@ Fiware-Correlator: REGEX([0-9a-f\-]{36})
 =======================================
 POST http://localhost:REGEX(\d+)/v1/queryContext
 Content-Length: 137
+Fiware-Servicepath: /
+Fiware-Servicepath: /
```
Comment (fgalan) [inline test/functionalTest/cases/0715_forward_xauth_token_to_cprs/propagate_xauth_token_for_update_and_query_standard_ops.test]: Dup

File: test/functionalTest/cases/0715_forward_xauth_token_to_cprs/propagate_xauth_token_for_update_and_query_standard_ops.test [hunk 8]
```diff
@@ -292,6 +296,8 @@ Fiware-Correlator: REGEX([0-9a-f\-]{36})
 }
 =======================================
 POST http://localhost:REGEX(\d+)/v1/updateContext
+Fiware-Servicepath: /
```
Comment (fgalan) [inline test/functionalTest/cases/0715_forward_xauth_token_to_cprs/propagate_xauth_token_for_update_and_query_standard_ops.test]: Dup

File: test/functionalTest/cases/0715_forward_xauth_token_to_cprs/propagate_xauth_token_for_update_and_query_standard_ops.test [hunk 9]
```diff
@@ -318,6 +324,8 @@ Fiware-Correlator: REGEX([0-9a-f\-]{36})
 }
 =======================================
 POST http://localhost:REGEX(\d+)/v1/queryContext
+Fiware-Servicepath: /
```
Comment (fgalan) [inline test/functionalTest/cases/0715_forward_xauth_token_to_cprs/propagate_xauth_token_for_update_and_query_standard_ops.test]: Dup

File: test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test [hunk 10]
```diff
@@ -270,6 +271,7 @@ Fiware-Correlator: REGEX([0-9a-f\-]{36})
 =======================================
 POST http://127.0.0.1:REGEX(\d+)/notify
 Fiware-Servicepath: /
+Fiware-Servicepath: /
```
Comment (fgalan) [inline test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test]: Dup

File: test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test [hunk 11]
```diff
@@ -295,6 +297,7 @@ Fiware-Correlator: REGEX([0-9a-f\-]{36})
...
```
Comment (fgalan) [inline test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test]: Dup

File: test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test [hunk 12]
```diff
@@ -320,6 +323,7 @@ Fiware-Correlator: REGEX([0-9a-f\-]{36})
...
```
Comment (fgalan) [inline test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test]: Dup

File: test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test [hunk 13]
```diff
@@ -346,6 +350,7 @@ Fiware-Correlator: REGEX([0-9a-f\-]{36})
 PUT http://127.0.0.1:REGEX(\d+)/notify?a1=true&type=T1&id=E1
 Fiware-Servicepath: /
 Entity-Id: E1
+Fiware-Servicepath: /
```
Comment (fgalan) [inline test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test]: Dup

File: test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test [hunk 14]
```diff
@@ -365,6 +370,7 @@ Fiware-Correlator: REGEX([0-9a-f\-]{36})
 PUT http://127.0.0.1:REGEX(\d+)/notify?a2=null&type=T2&id=E1
 Fiware-Servicepath: /
 Entity-Id: E1
+Fiware-Servicepath: /
```
Comment (fgalan) [inline test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test]: Dup

File: test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test [hunk 15]
```diff
@@ -385,6 +391,7 @@ PUT http://127.0.0.1:REGEX(\d+)/notify?a3=3&type=T3&id=E1
 Fiware-Servicepath: /
 Ngsiv2-Attrsformat: custom
 Entity-Id: E1
+Fiware-Servicepath: /
```
Comment (fgalan) [inline test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test]: Dup

File: test/functionalTest/cases/2584_service_path_header_in_notifs/service_path_header_in_notifs.test [hunk 16]
```diff
@@ -0,0 +1,196 @@
+# Copyright 2017 Telefonica Investigacion y Desarrollo, S.A.U
```
Comment (fgalan) [inline test/functionalTest/cases/2584_service_path_header_in_notifs/service_path_header_in_notifs.test]: Do we really need a new .test for this when the large ammount of notifications in default SP we have in another tests will ensure the functionality is correct? I does't suggest to remove it (as it is already written) just a remark to think twice next time in similar situations ;)

NTC

Comment (kzangeli) [inline test/functionalTest/cases/2584_service_path_header_in_notifs/service_path_header_in_notifs.test]: This is merely a tool that I needed while trying to fix the bug.
And you're right, once the bug is fixed, no need for this test anymore.

And to anticipate :-):
  No, it is not easier to try to find a 'half valid' already existing functest than to create a new one.
  A new one is both faster and better suited for the problem
  To create  a functest takes me a few minutes, copying from older functests (as I believe we all do)

File: test/functionalTest/cases/0673_xauth_token/auth_token_from_subscribe_request.test (hunk 1 above)
Comment (kzangeli) [inline test/functionalTest/cases/0673_xauth_token/auth_token_from_subscribe_request.test]: Fixed in deeabf3

File: test/functionalTest/cases/0673_xauth_token/auth_token_from_subscribe_request.test (hunk 2 above)
Comment (kzangeli) [inline test/functionalTest/cases/0673_xauth_token/auth_token_from_subscribe_request.test]: Fixed in deeabf3

File: test/functionalTest/cases/0673_xauth_token/auth_token_three_steps.test (hunk 3 above)
Comment (kzangeli) [inline test/functionalTest/cases/0673_xauth_token/auth_token_three_steps.test]: Fixed in deeabf3

File: test/functionalTest/cases/0673_xauth_token/auth_token_three_steps.test (hunk 4 above)
Comment (kzangeli) [inline test/functionalTest/cases/0673_xauth_token/auth_token_three_steps.test]: Fixed in deeabf3

File: test/functionalTest/cases/0673_xauth_token/subscription.test (hunk 5 above)
Comment (kzangeli) [inline test/functionalTest/cases/0673_xauth_token/subscription.test]: Fixed in deeabf3

File: test/functionalTest/cases/0715_forward_xauth_token_to_cprs/propagate_xauth_token_for_update_and_query_standard_ops.test (hunk 6 above)
Comment (kzangeli) [inline test/functionalTest/cases/0715_forward_xauth_token_to_cprs/propagate_xauth_token_for_update_and_query_standard_ops.test]: Fixed in deeabf3

File: test/functionalTest/cases/0715_forward_xauth_token_to_cprs/propagate_xauth_token_for_update_and_query_standard_ops.test (hunk 7 above)
Comment (kzangeli) [inline test/functionalTest/cases/0715_forward_xauth_token_to_cprs/propagate_xauth_token_for_update_and_query_standard_ops.test]: Fixed in deeabf3

File: test/functionalTest/cases/0715_forward_xauth_token_to_cprs/propagate_xauth_token_for_update_and_query_standard_ops.test (hunk 8 above)
Comment (kzangeli) [inline test/functionalTest/cases/0715_forward_xauth_token_to_cprs/propagate_xauth_token_for_update_and_query_standard_ops.test]: Fixed in deeabf3

File: test/functionalTest/cases/0715_forward_xauth_token_to_cprs/propagate_xauth_token_for_update_and_query_standard_ops.test (hunk 9 above)
Comment (kzangeli) [inline test/functionalTest/cases/0715_forward_xauth_token_to_cprs/propagate_xauth_token_for_update_and_query_standard_ops.test]: Fixed in deeabf3

File: test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test (hunk 10 above)
Comment (kzangeli) [inline test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test]: Fixed in deeabf3

File: test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test (hunk 11 above)
Comment (kzangeli) [inline test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test]: Fixed in deeabf3

File: test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test (hunk 12 above)
Comment (kzangeli) [inline test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test]: Fixed in deeabf3

File: test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test (hunk 13 above)
Comment (kzangeli) [inline test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test]: Fixed in deeabf3

File: test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test (hunk 14 above)
Comment (kzangeli) [inline test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test]: Fixed in deeabf3

File: test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test (hunk 15 above)
Comment (kzangeli) [inline test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test]: Fixed in deeabf3

File: test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test (hunk 10 above)
Comment (fgalan) [inline test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test]: This is an example of dup with an already existing "Fiware-Servicpath: /" line

File: test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test (hunk 11 above)
Comment (fgalan) [inline test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test]: This is an example of dup with an already existing "Fiware-Servicpath: /" line

File: test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test (hunk 12 above)
Comment (fgalan) [inline test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test]: This is an example of dup with an already existing "Fiware-Servicpath: /" line

File: test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test (hunk 13 above)
Comment (fgalan) [inline test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test]: This is an example of dup with an already existing "Fiware-Servicpath: /" line

File: test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test (hunk 14 above)
Comment (fgalan) [inline test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test]: This is an example of dup with an already existing "Fiware-Servicpath: /" line

File: test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test (hunk 15 above)
Comment (fgalan) [inline test/functionalTest/cases/2015_notification_templates/notification_templates_many_notifications.test]: This is an example of dup with an already existing "Fiware-Servicpath: /" line

File: test/functionalTest/cases/2584_service_path_header_in_notifs/service_path_header_in_notifs.test (hunk 16 above)
Comment (fgalan) [inline test/functionalTest/cases/2584_service_path_header_in_notifs/service_path_header_in_notifs.test]: > A new one is both faster and better suited for the problem

Faster and better in the context of a given isolated problem solution task, but not so fast and good in the overall. Each new .test is another lilttle piece of stuff to mantain (and when you have to do a large refactor effort, e.g. XML removal, these little pieces, all together, have a significative cost). In addition, each new .test is a little more delay in make test, so impacting our development speed (even in the case of "I'm working on the next development while make test ends" part of your mind is still in the PR that you haven't merged yet due to the make test... at least in my case I cannot do complete context change as computers do ;).

It isnt' a golde rule about this, but something to think about (food for thought).

**Comment by fgalan** [comment]
LGTM

**Pull request merged**